### PR TITLE
Derivative doit() fixed

### PR DESCRIPTION
**Pull request by vishalg2235**
Fixes `Derivative(f(x), (x, n)).doit()` in https://github.com/sympy/sympy/issues/13843
For `n` the value of `self.variables = ()` which is removing `n` from Derivative. `self.variable_count = (x, n)` and hence returns correct Derivative.
```
>>> from sympy import *
>>> m, n = symbols('m n', integer=True)
>>> x = symbols('x')
>>> f = Function('f')
>>> Derivative(f(x), (x, n)).doit()
Derivative(f(x), (x, n))
```

`Derivative(Derivative(f(x), (x, m)), (x, n))` I'am still fixing, following is changes that I have done but its still not solved! Please have a look.
```
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -561,7 +561,10 @@ def _eval_derivative(self, s):
         l = []
         for a in self.args:
             i += 1
-            da = a.diff(s)
+            if type(s) == tuple:
+                da = a.diff(s[0])
+            else:
+                da = a.diff(s)
             if da is S.Zero:
                 continue
             try:
@@ -1104,7 +1107,8 @@ def __new__(cls, expr, *variables, **assumptions):
         from sympy.tensor.array import Array, NDimArray, derive_by_array
 
         expr = sympify(expr)
-
+        import pdb
+        pdb.set_trace()
         # There are no variables, we differentiate wrt all of the free symbols
         # in expr.
         if not variables:
@@ -1235,7 +1239,7 @@ def __new__(cls, expr, *variables, **assumptions):
 
             if unhandled_non_symbol:
                 obj = None
-            elif not count.is_Integer:
+            elif not count.is_Integer and not count.is_integer:
                 obj = None
             else:
                 if isinstance(v, (collections.Iterable, Tuple, MatrixCommon, NDimArray)):
@@ -1249,11 +1253,16 @@ def __new__(cls, expr, *variables, **assumptions):
                     old_v = v
                     v = new_v
                 obj = expr
-                for i in range(count):
-                    obj2 = deriv_fun(obj, v)
-                    if obj == obj2:
-                        break
-                    obj = obj2
+                pdb.set_trace()
+                try:
+                    for i in range(count):
+                        obj2 = deriv_fun(obj, v)
+                        if obj == obj2:
+                            break
+                        obj = obj2
+                        nderivs += 1
+                except TypeError:
+                    obj = deriv_fun(obj, Tuple(v, count))
                     nderivs += 1
                 if not is_symbol:
                     if obj is not None:
@@ -1414,7 +1423,7 @@ def doit(self, **hints):
         if hints.get('deep', True):
             expr = expr.doit(**hints)
         hints['evaluate'] = True
-        return self.func(expr, *self.variables, **hints)
+        return self.func(expr, *self.variable_count, **hints)
 
     @_sympifyit('z0', NotImplementedError)
     def doit_numerically(self, z0):
```

### CODE REVIEW

**Comment by normalhuman** [comment]
The issue with combining is that the present algorithm is based on `for i in range(count):` and that can't work if `count` is a symbol. Entirely new logic for combining (x, n) should be written, and it should be in the branch `elif not count.is_Integer:` which is presently simply `obj=None` (nothing is done). This logic should involve comparing the current `variable` to other variables in Derivative, and paying attention to commutativity. As noted in comments, 

  * Derivative wrt different symbols commute.
  * Derivative wrt different non-symbols commute.
  * Derivatives wrt symbols and non-symbols don't commute.

**Comment by vishalg2235** [comment]
Yes. If count is symbol then `range(count)` raises TypeError. That's why I have used try-except. 
For `count=n `, `n.is_integer=True` and  `n.is_Integer=False` so I think if it's a symbol then after `if n.is_integer` new logic for combining (x, n) should be placed.

**Comment by vishalg2235** [comment]
@normalhuman @asmeurer @Upabjojr Now derivatives are commutative. Please review!
now: 
```
>>> from sympy import *
>>> m, n = symbols('m n', integer=True)
>>> x = symbols('x')
>>> f = Function('f')
>>> Derivative(Derivative(f(x), (x, m)), (x, n))
Derivative(f(x), (x, m + n))
>>> Derivative(Derivative(f(x), (x, m)), (x, 3))
Derivative(f(x), (x, m + 3))
>>> Derivative(Derivative(f(x), (x, 5)), (x, n))
Derivative(f(x), (x, n + 5))
>>> Derivative(Derivative(f(x), (x, m+5)), (x, n))
Derivative(f(x), (x, m + n + 5))
>>> Derivative(Derivative(f(x), (x, m+5)), (x, n+3))
Derivative(f(x), (x, m + n + 8))
```

File: sympy/core/function.py [hunk 1]
```diff
@@ -1249,11 +1249,15 @@ def __new__(cls, expr, *variables, **assumptions):
                     old_v = v
                     v = new_v
                 obj = expr
-                for i in range(count):
-                    obj2 = deriv_fun(obj, v)
-                    if obj == obj2:
-                        break
-                    obj = obj2
+                try:
```
Comment (Upabjojr) [inline sympy/core/function.py]: Is it possible to handle the problem without `try`? It usually makes the code complicated to read.

File: sympy/core/tests/test_function.py [hunk 2]
```diff
@@ -859,6 +859,13 @@ def test_issue_12005():
     assert e5.diff(x) == Derivative(f(x), x, x)
     assert f(g(x)).diff(g(x), g(x)) == Subs(Derivative(f(y), y, y), (y,), (g(x),))
 
+def test_issue_13843():
+    x = symbols('x')
+    f = Function('f')
+    m, n = symbols('m n', integer=True)
```
Comment (Upabjojr) [inline sympy/core/tests/test_function.py]: What if you remove the `integer=True` condition?

File: sympy/core/function.py (hunk 1 above)
Comment (vishalg2235) [reply sympy/core/function.py]: Should I use `if-else` statement ?

File: sympy/core/tests/test_function.py (hunk 2 above)
Comment (vishalg2235) [reply sympy/core/tests/test_function.py]:
```
>>> m, n = symbols('m n')
>>> Derivative(Derivative(f(x), (x, m+5)), (x, n+3))
Derivative(Derivative(f(x), (x, m + 5)), (x, n + 3))
```

**Comment by vishalg2235** [comment]
It is failing [this](https://github.com/sympy/sympy/blob/master/sympy/core/tests/test_diff.py#L118) test. There is some problem occurring in `subs`
```
>>> x = symbols('x')
>>> m, n = symbols('m n', integers=True)
>>> expr = diff(sin(x),(x,n))
>>> expr
Derivative(sin(x), (x, n))
>>> expr.subs(sin(x),cos(-x))    # This should be: Derivative(cos(-x), (x, n))
Derivative(cos(x), (x, n))
```

File: sympy/core/function.py (hunk 1 above)
Comment (Upabjojr) [reply sympy/core/function.py]: I guess the exception is raised if `count` is a symbol or non-number.

File: sympy/core/function.py [hunk 3]
```diff
@@ -1235,7 +1235,7 @@ def __new__(cls, expr, *variables, **assumptions):
 
             if unhandled_non_symbol:
                 obj = None
-            elif not count.is_Integer:
+            elif not count.is_Integer and not count.is_integer:
```
Comment (Upabjojr) [inline sympy/core/function.py]: I'm not sure about this edit either. This should be the case where `count` is symbolic.

File: sympy/core/tests/test_function.py [hunk 4]
```diff
@@ -859,6 +859,13 @@ def test_issue_12005():
...
+    assert Derivative(Derivative(f(x), (x, m)), (x, n)) == Derivative(f(x), (x, m + n))
+    assert Derivative(Derivative(f(x), (x, m+5)), (x, n+3)) == Derivative(f(x), (x, m + n + 8))
```
Comment (Upabjojr) [inline sympy/core/tests/test_function.py]: Missing the test for `.doit()`.

**Comment by Upabjojr** [comment]
I pushed some edits. The code was otherwise too messy.

**Comment by vishalg2235** [comment]
Thanks @Upabjojr now code is even more understanding to me.

**Comment by Upabjojr** [comment]
Merged, thanks for your contribution!

**Pull request merged**